exchanges/ftx/ftx_websocket.go

```diff
@@ -0,0 +1,523 @@
+// WsAuth sends an authentication message to receive auth data
+func (f *FTX) WsAuth() error {
+	intNonce := time.Now().UnixNano() / 1000000
+	strNonce := strconv.FormatInt(intNonce, 10)
+	hmac := crypto.GetHMAC(
+		crypto.HashSHA256,
+		[]byte(strNonce+"websocket_login"),
+		[]byte(f.API.Credentials.Secret),
+	)
+	sign := crypto.HexEncodeToString(hmac)
+	req := Authenticate{Operation: "login",
+		Args: AuthenticationData{
+			Key:  f.API.Credentials.Key,
+			Sign: sign,
+			Time: intNonce,
+		},
+	}
+	return f.WebsocketConn.SendJSONMessage(req)
+}
+
+// Subscribe sends a websocket message to receive data from the channel
+func (f *FTX) Subscribe(channelToSubscribe wshandler.WebsocketChannelSubscription) error {
+	var sub WsSub
+	switch channelToSubscribe.Channel {
+	case wsFills, wsOrders, wsMarkets:
+		sub.Operation = subscribe
+		sub.Channel = channelToSubscribe.Channel
+	default:
+		a, err := f.GetPairAssetType(channelToSubscribe.Currency)
+		if err != nil {
+			return err
+		}
+		sub.Operation = subscribe
+		sub.Channel = channelToSubscribe.Channel
+		sub.Market = f.FormatExchangeCurrency(channelToSubscribe.Currency, a).String()
+	}
+	return f.WebsocketConn.SendJSONMessage(sub)
+}
+
+// GenerateDefaultSubscriptions generates default subscription
+func (f *FTX) GenerateDefaultSubscriptions() {
+	var subscriptions []wshandler.WebsocketChannelSubscription
+	subscriptions = append(subscriptions, wshandler.WebsocketChannelSubscription{
+		Channel: wsMarkets,
+	})
+	var channels = []string{wsTicker, wsTrades, wsOrderbook}
+	for a := range f.CurrencyPairs.AssetTypes {
+		pairs := f.GetEnabledPairs(f.CurrencyPairs.AssetTypes[a])
+		for z := range pairs {
+			newPair := currency.NewPairWithDelimiter(pairs[z].Base.String(), pairs[z].Quote.String(), "-")
+			for x := range channels {
+				subscriptions = append(subscriptions, wshandler.WebsocketChannelSubscription{
+					Channel:  channels[x],
+					Currency: newPair,
+				})
+			}
+		}
+	}
+	f.Websocket.SubscribeToChannels(subscriptions)
+}
+
+// GenerateAuthSubscriptions generates default subscription
+func (f *FTX) GenerateAuthSubscriptions() {
+	var subscriptions []wshandler.WebsocketChannelSubscription
+	var channels = []string{wsOrders, wsFills}
+	for x := range channels {
+		subscriptions = append(subscriptions, wshandler.WebsocketChannelSubscription{
+			Channel: channels[x],
+		})
+	}
+	f.Websocket.SubscribeToChannels(subscriptions)
+}
+
+// wsReadData gets and passes on websocket messages for processing
+func (f *FTX) wsReadData() {
+	f.Websocket.Wg.Add(1)
+
+	defer f.Websocket.Wg.Done()
+
+	for {
+		select {
+		case <-f.Websocket.ShutdownC:
+			return
+
+		default:
+			resp, err := f.WebsocketConn.ReadMessage()
+			if err != nil {
+				f.Websocket.ReadMessageErrors <- err
+				return
+			}
+			f.Websocket.TrafficAlert <- struct{}{}
+			err = f.wsHandleData(resp.Raw)
+			if err != nil {
+				f.Websocket.DataHandler <- err
+			}
+		}
+	}
+}
+
+func timestampFromFloat64(ts float64) time.Time {
```

Comment:
Please add a test for this and include an example from one of the websocket updates / API docs